contributors/devel/sig-architecture/feature-gates.md

```diff
@@ -1,87 +1,256 @@
+Very rarely we will deprecate some user-visible aspect of Kubernetes (almost
+always something that has no actual impact or which is truly a bug). When we
+do that, we also use a feature gate. The pattern is to name the gate so it
+describes the functionality being deprecated and to default the value to
+`false`.
+
+If some user is impacted by the deprecation, they can set that gate to `true`
+to unbreak themselves (and then file a bug). If this happens, we must
+reconsider the deprecation and may choose to abandon it entirely by changing
+the gate back to `true` for a release or two and eventually removing it.
+
+NOTE: We do not generally remove fields from the API.
+
+### Other scenarios
+
+Sometimes we use `{Default: true, PreRelease: featuregate.Beta}` for keeping
+legacy behavior on while a new alternative is being implemented. When the new
+default behavior is transitioning to GA and it is time to drop the old legacy
+behavior, we will end up with `{Default: false, PreRelease: featuregate.GA,
+LockToDefault: true}`. For an example, see: [LegacyNodeRoleBehavior &
+ServiceNodeExclusion](https://github.com/kubernetes/kubernetes/pull/97543/files).
+
+## Using a feature gate in code
 
-## Other scenarios
+Feature gates are basically fancy `bool` variables. They are either enabled or
+disabled. When implementing a feature gate, there are a few patterns you might
+follow.
+
+### Features which add a new API field
+
+All features which add API fields *must* start in alpha. This ensures that once
+the feature becomes beta (and the fields get used) it is possible to roll-back
+one Kubernetes version and not have the data be discarded. Users who enable and
+use alpha features (which are disabled by default) do not have a safe rollback
+target.
+
+When a feature gate is disabled, the system should behave as if the API fields
+do not exist. API operations which try to use the field are expected to proceed
+as if the field was unknown and the "extra" data was discarded.
+
+The API's registry code (`pkg/registry/...`) must check the gate before
+validation. If the gate is disabled and the operation is a CREATE, the new
+field must be removed (set to `nil`). If the gate is disabled and the
+operation is an UPDATE, the previous form of the object must be checked. Only
+if this object was not already using this field should it be removed. This
+usually manifests as something like:
+
+```
+if disabled(gate) && !newFieldInUse(oldObj) {
+    obj.newField = nil
+}
+```
+
+Once this is done, API validation should *not* check the gate - if the field is
+carrying a value, it should be validated. This ensures that once an object is
```

Comment:
this comment is hard to understand , at least for me as non native, I think that is the `-` to connect with the other sentence, this is some way I understand it better
```suggestion
Once this is done, API validation should *not* check the gate EXCEPT IF the field is
carrying a value, THEN it should be validated. This ensures that once an object is
```

Reply:
I will clarify.